integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateUtil.scala

```diff
@@ -166,127 +208,160 @@ object PreAggregateUtil {
+          newColumnName,
+          max.prettyName)
+      case Average(MatchCastExpression(exp: Expression, changeDataType: DataType)) =>
+        list += getFieldForAggregateExpression(exp,
           changeDataType,
-          "count",
-          carbonTable.getColumnByName(parentTableName, attr.name).getColumnId,
-          parentTableName,
-          parentDatabaseName, parentTableId = parentTableId)
+          carbonTable,
+          newColumnName,
+          "sum")
+        list += getFieldForAggregateExpression(exp,
+          changeDataType,
+          carbonTable,
+          newColumnName,
+          "count")
+      case avg@Average(exp: Expression) =>
+        list += getFieldForAggregateExpression(exp,
+          avg.dataType,
+          carbonTable,
+          newColumnName,
+          "sum")
+        list += getFieldForAggregateExpression(exp,
+          avg.dataType,
+          carbonTable,
+          newColumnName,
+          "count")
       case others@_ =>
         throw new MalformedCarbonCommandException(s"Un-Supported Aggregation Type: ${
           others.prettyName}")
     }
   }
 
+  /**
+   * Below method will be used to get the field and its data map field object
+   * for aggregate expression
+   * @param expression
+   *                   expression in aggregate function
+   * @param dataType
+   *                 data type
+   * @param carbonTable
+   *                    parent carbon table
+   * @param newColumnName
+   *                      column name of aggregate table
+   * @param aggregationName
+   *                        aggregate function name
+   * @return field and its metadata tuple
+   */
+  def getFieldForAggregateExpression(expression: Expression,
+      dataType: DataType,
+      carbonTable: CarbonTable,
+      newColumnName: String,
+      aggregationName: String): (Field, DataMapField) = {
+    val parentColumnsName = new ArrayBuffer[String]()
+    expression.transform {
+      case attr: AttributeReference =>
+        parentColumnsName += attr.name
+        attr
+    }
+    val arrayBuffer = new ArrayBuffer[ColumnTableRelation]()
+    parentColumnsName.foreach { name =>
+      arrayBuffer += getColumnRelation(name,
+        carbonTable.getAbsoluteTableIdentifier.getCarbonTableIdentifier.getTableId,
+        carbonTable.getAbsoluteTableIdentifier.getCarbonTableIdentifier.getTableName,
+        carbonTable.getAbsoluteTableIdentifier.getCarbonTableIdentifier.getDatabaseName,
+        carbonTable)
+    }
+    // if parent column relation is of size more than one that means aggregate table
+    // column is derived from multiple column of main table
+    // or if expression is not a instance of attribute reference
+    // then use column name which is passed
+    val columnName =
+    if (parentColumnsName.size > 1 && !expression.isInstanceOf[AttributeReference]) {
+      newColumnName
+    } else {
+      expression.asInstanceOf[AttributeReference].name
+    }
+    getField(columnName,
```

Comment:
Can you rename `getField` to `createField` since it is creating a new Field object
Rename other similar function also, like `getFieldForAggregateExpression`